File: packages/tactical-engine-py/tactical_engine/movement/path_cost.py
from shared_contracts import Coordinate


def is_diagonal_step(from_: Coordinate, to: Coordinate) -> bool:
    return abs(from_.x - to.x) == 1 and abs(from_.y - to.y) == 1


def step_cost(from_: Coordinate, to: Coordinate, diagonal_index: int) -> int:
    if not is_diagonal_step(from_, to):
        return 5
    # D&D 5e alternating diagonal rule: 1st diagonal = 5ft, 2nd = 10ft, repeating
    return 5 if diagonal_index % 2 == 0 else 10


def compute_path_cost(path: list[Coordinate]) -> int:
    total = 0
    diagonal_index = 0
    for i in range(1, len(path)):
        cost = step_cost(path[i - 1], path[i], diagonal_index)
        total += cost
        if is_diagonal_step(path[i - 1], path[i]):
            diagonal_index += 1
    return total

File: packages/tactical-engine/src/movement/path-cost.js
export function isDiagonalStep(from, to) {
    return Math.abs(from.x - to.x) === 1 && Math.abs(from.y - to.y) === 1;
}
export function stepCost(from, to, diagonalIndex, cellMultiplier = 1) {
    const baseCost = isDiagonalStep(from, to) ? (diagonalIndex % 2 === 0 ? 5 : 10) : 5;
    return baseCost * cellMultiplier;
}
export function computePathCost(path, getMultiplier) {
    let total = 0;
    let diagonalIndex = 0;
    for (let index = 1; index < path.length; index += 1) {
        const dest = path[index];
        const multiplier = getMultiplier ? getMultiplier(dest) : 1;
        const cost = stepCost(path[index - 1], dest, diagonalIndex, multiplier);
        total += cost;
        if (isDiagonalStep(path[index - 1], dest)) {
            diagonalIndex += 1;
        }
    }
    return total;
}
